### PR TITLE
[FSDP]Skip unshard call during checkpointing for NO_SHARD sharding strategy

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101095

### CODE REVIEW

**Review by fegin — APPROVED** [review]
LGTM, thanks for the fix

**Comment by fegin** [comment]
Probably need to resubmit to trigger the CI, there was some infra issue.

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 1]
```diff
@@ -344,7 +363,11 @@ def _full_pre_load_state_dict_hook(
 def _full_post_load_state_dict_hook(
     module: nn.Module, fsdp_state: _FSDPState, *args, **kwargs
 ) -> None:
-    _exit_unshard_params_ctx(module, fsdp_state)
+    if not (
+        _is_composable(fsdp_state)
+        and fsdp_state.sharding_strategy == ShardingStrategy.NO_SHARD
+    ):
+        _exit_unshard_params_ctx(module, fsdp_state)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_state_dict_utils.py:370]: wonder if duplication of these chekcs can be reduced by putting it inside exit / enter unshard params ctx?

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 2]
```diff
@@ -146,6 +146,12 @@ def _common_unshard_pre_state_dict_hook(
     Performs the pre-state_dict tasks shared by all state_dict types that require
     ``_unshard_fsdp_state_params()``. FULL_STATE_DICT and SHARDED_STATE_DICT use this hook.
     """
+    # For composable `fully_shard`, it does not need to unshard parameters for `NO_SHARD` cases.
+    if (
+        _is_composable(fsdp_state)
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/fsdp/_state_dict_utils.py:151]: I have no idea :/

**Pull request closed without merging**